### PR TITLE
741 permissions refactoring

### PR DESCRIPTION
**Pull request by SofiaSazonova**
### Feature or Bugfix

- Refactoring

### Detail
- core.permissions refactoring
- separate Service and Repository logic
- separate TenantPermissions, Resource policy-Permissions and Permissions
- move decorators to separate folder

### Relates
- [741](https://github.com/data-dot-all/dataall/issues/741)

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?N/A
 
- Does this PR introduce any functionality or component that requires authorization? N/A
  
- Are you using or adding any cryptographic features? N/A
 
- Are you introducing any new policies/roles/users? N/A

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by dlpzx** [comment]
Regarding the code layout for the permissions module, here are some suggestions:
- `constants/permissions` are basically the core_permissions = environment_permissions, group_permissions, organization_permissions. I think we should include them not as constants but as a part of `services` like we do in the module (e.g. `dataall/modules/dashboards/services/dashboard_permissions.py`). We can do this in the permissions module `permissions/services/core_permissions.py` or in each of the modules: `environment/services/environment_permissions.py` ....
- A second idea (in which I am less opinionated), is to include the `dataall/core/permissions/decorators/permission_checker.py` not in a new layer, but in the `services` layer. Just to keep it simple and not include new layers/components in the design

File: backend/dataall/core/permissions/api/resolvers.py
Comment (dlpzx) [inline backend/dataall/core/permissions/api/resolvers.py:6]: Unused import statement in line 6

File: backend/dataall/core/permissions/services/tenant_policy_service.py
Comment (dlpzx) [inline backend/dataall/core/permissions/services/tenant_policy_service.py:2]: Instead of importing all permissions, we could import only the one that is used TENANT_ALL (Look at suggestion in overall comments)

File: backend/dataall/core/permissions/services/tenant_policy_service.py
Comment (dlpzx) [inline backend/dataall/core/permissions/services/tenant_policy_service.py:6]: Instead renaming the TenantPolicy as TenantPolicyModel, we could rename the TenantPolicy coming fromt enant_policy_repositories as TenantPolicyRepository to keep the same naming convention that we have followed in the rest of the modules

File: backend/dataall/core/permissions/services/tenant_policy_service.py
Comment (dlpzx) [inline backend/dataall/core/permissions/services/tenant_policy_service.py:51]: We can merge `validate_add_permission_to_tenant_policy_params` and `validate_add_permissions_params` into the same function right?

File: backend/dataall/core/permissions/services/tenant_policy_service.py
Comment (dlpzx) [reply backend/dataall/core/permissions/services/tenant_policy_service.py:10]: Although I would make a small change to differentiate the methods that validate input parameters (e.g. `validate_find_tenant_policy`) and those that validate the permissions of the user (e.g. `validate_admin_access`). 
We have 2 options. First we could clearly define the function names so that it is clear the type of validation. For the above case, `validate_find_tenant_policy` ---> `validate_find_tenant_policy_params`. Or option 2 (which I think I prefer) is to split the `TenantPolicyValidationService` into 2 separate classes: `RequestValidator` with the validation of the input parameters and `TenantPolicyValidationService` with the service that validates user permissions. What are your thoughts?

File: backend/dataall/core/permissions/services/tenant_policy_service.py
Comment (dlpzx) [inline backend/dataall/core/permissions/services/tenant_policy_service.py:59]: TenantPolicy.TENANT_NAME is only used in `dataall/core/permissions/services/tenant_policy_service.py` and seems more of a TenantPolicyService variable than a `dataall/core/permissions/db/tenant/tenant_policy_repositories.py` constant. I think we could move it to the service

File: backend/dataall/core/permissions/services/tenant_policy_service.py
Comment (dlpzx) [inline backend/dataall/core/permissions/services/tenant_policy_service.py:135]: `if not groups...` should be part of `RequestValidator`

**Review by dlpzx — COMMENTED** [review]
Halfway reviewed!

File: backend/dataall/core/permissions/services/tenant_policy_service.py
Comment (dlpzx) [inline backend/dataall/core/permissions/services/tenant_policy_service.py:228]: if not groups... should be part of RequestValidator

File: backend/dataall/core/permissions/services/tenant_policy_service.py
Comment (dlpzx) [inline backend/dataall/core/permissions/services/tenant_policy_service.py:243]: if not group_uri and tenant_name... should be part of RequestValidator

File: backend/dataall/core/permissions/services/tenant_policy_service.py
Comment (dlpzx) [inline backend/dataall/core/permissions/services/tenant_policy_service.py:66]: The name of this function is too generic for being used just once in a very particular case (action: `UPDATE_TENANT_TEAM_PERMISSIONS`) Either we make the function more generic or we rename it to be more specific

File: backend/dataall/core/permissions/db/tenant/tenant_repositories.py
Comment (dlpzx) [inline backend/dataall/core/permissions/db/tenant/tenant_repositories.py:4]: In all the functions of `Tenant` we should not be validating the input (`if note tenant_name....`). That should happen in the `api/resolver`

File: backend/dataall/core/permissions/db/tenant/tenant_policy_repositories.py
Comment (dlpzx) [inline backend/dataall/core/permissions/db/tenant/tenant_policy_repositories.py:5]: Remove unused import statements line 5,7 10

File: backend/dataall/core/permissions/db/tenant/tenant_policy_repositories.py
Comment (dlpzx) [inline backend/dataall/core/permissions/db/tenant/tenant_policy_repositories.py:16]: Already commented before, `TENANT_NAME` should be part of service

Comment (dlpzx) [inline backend/dataall/core/permissions/db/tenant/tenant_policy_repositories.py:16]: Define `DAAdministrators` as constant instead of hardcoding it in line 102

File: backend/dataall/core/permissions/db/tenant/tenant_policy_repositories.py
Comment (dlpzx) [inline backend/dataall/core/permissions/db/tenant/tenant_policy_repositories.py:22]: Input validation should happen in `api/resolver`

File: backend/local_graphql_server.py
Comment (dlpzx) [inline backend/local_graphql_server.py:75]: Instead of hardcoding `dataall` we could use `TENANT_NAME = 'dataall'` from `TenantPolicyService`  currently it is defined in repositories, but it might be better placed in service

File: backend/dataall/core/permissions/db/resource_policy/resource_policy_models.py
Comment (dlpzx) [inline backend/dataall/core/permissions/db/resource_policy/resource_policy_models.py:3]: Remove unused import statement line 3

File: backend/dataall/core/permissions/db/resource_policy/resource_policy_repositories.py
Comment (dlpzx) [inline backend/dataall/core/permissions/db/resource_policy/resource_policy_repositories.py:11]: Rename `ResourcePolicy` to `ResourcePolicyRepository`

File: backend/dataall/core/permissions/db/resource_policy/resource_policy_repositories.py
Comment (dlpzx) [inline backend/dataall/core/permissions/db/resource_policy/resource_policy_repositories.py:6]: Rename to `PermissionRepository`

Comment (dlpzx) [inline backend/dataall/core/permissions/db/resource_policy/resource_policy_repositories.py:6]: `from dataall.core.permissions.db.permission.permission_models import Permission`

File: backend/dataall/core/permissions/db/resource_policy/resource_policy_repositories.py
Comment (dlpzx) [inline backend/dataall/core/permissions/db/resource_policy/resource_policy_repositories.py:9]: `from dataall.core.permissions.db.resource_policy.resource_policy_models import ResourcePolicy, ResourcePolicyPermission`

File: backend/dataall/core/permissions/db/resource_policy/resource_policy_repositories.py
Comment (dlpzx) [inline backend/dataall/core/permissions/db/resource_policy/resource_policy_repositories.py:13]: Looking at this db repository class, it contains a lot of business logic. Maybe we can add a `resource_policy_service.py` service and keep the db package limited to db operations.

Comment (dlpzx) [reply backend/dataall/core/permissions/db/resource_policy/resource_policy_repositories.py:13]: For example, methods such as: `check_user_resource_permission`, `update_resource_policy`, `attach_resource_policy`... are not db related, they are purely business logic

File: backend/dataall/core/permissions/db/permission/permission_repositories.py
Comment (dlpzx) [inline backend/dataall/core/permissions/db/permission/permission_repositories.py:9]: Similar comments than the ones in Resource_Policy: 
1. we should rename this to PermissionsRepository
2. all the business logic should be placed in the services layer, so maybe we need to create a `permissions_service.py` that contains functions such as `get_permission_by_name`...

File: backend/dataall/core/permissions/db/group/group_policy_repositories.py
Comment (dlpzx) [inline backend/dataall/core/permissions/db/group/group_policy_repositories.py:5]: Rename to `GroupPolicyRepository` and move business logic to `services`

**Review by dlpzx — CHANGES_REQUESTED** [review]
It is not 100% reviewed, but there are enough comments to work on. I will do a second review once those are addressed

File: backend/dataall/core/permissions/services/tenant_policy_service.py
Comment (SofiaSazonova) [reply backend/dataall/core/permissions/services/tenant_policy_service.py:10]: I like splitting in two. I'm totally totally the person, who has a bag with bags and a box with boxes))

File: backend/dataall/core/permissions/db/tenant/tenant_repositories.py
Comment (SofiaSazonova) [reply backend/dataall/core/permissions/db/tenant/tenant_repositories.py:4]: Yes, It's rudimental check. Everything is already checked in services.
As for "save_tenant" it's only used with 'dataall' constant;

File: backend/dataall/core/permissions/db/tenant/tenant_policy_repositories.py
Comment (SofiaSazonova) [reply backend/dataall/core/permissions/db/tenant/tenant_policy_repositories.py:16]: Good catch. Also, I used in in the TenantPolicyValidationService

**Comment by SofiaSazonova** [comment]
@dlpzx I made another huge step in refactoring) Take a look))

File: backend/local_graphql_server.py
Comment (dlpzx) [inline backend/local_graphql_server.py:11]: I still think that we should define the permissions in the service layer not in constants to keep it consistent to how we do it across the repo. Plus they are not constants, permissions are a central piece of the business logic

Comment (dlpzx) [reply backend/local_graphql_server.py:11]: In addition, i would import TENANT_ALL permissions only

File: backend/dataall/core/permissions/db/tenant/tenant_policy_repositories.py
Comment (dlpzx) [reply backend/dataall/core/permissions/db/tenant/tenant_policy_repositories.py]: This one is not fixed (line 42 and 43)

**Review by dlpzx — COMMENTED** [review]
.

File: backend/dataall/core/permissions/services/tenant_policy_service.py
Comment (dlpzx) [reply backend/dataall/core/permissions/services/tenant_policy_service.py]: matryoshka matrioska

File: backend/dataall/core/permissions/api/resolvers.py
Comment (petrkalos) [inline backend/dataall/core/permissions/api/resolvers.py:26]: `not filter` will evaluate all the [falsey values](https://www.geeksforgeeks.org/truthy-vs-falsy-values-in-python/#:~:text=Falsy%20Values%20Includes%3A) to true, hence this is not functionally equivalent. Do you see this as a problem here?

**Comment by petrkalos** [comment]
As a general comment, I'd split this in at least 2 PRs (maybe a third isn't possible)
1. Automatic rename
2. Move files around
3. Functional changes


By doing so I wouldn't even review n1 and n2

File: backend/dataall/core/permissions/api/resolvers.py
Comment (SofiaSazonova) [reply backend/dataall/core/permissions/api/resolvers.py:13]: It's rudimental check. Everything is validated inside `TenantPolicyService.update_group_permissions `, so it can be called not only from resolvers (if needed) safely.

**Comment by SofiaSazonova** [comment]
@petrkalos sorry, this one is already very mixed(
I will try to do it in next PRs

**Comment by SofiaSazonova** [comment]
@dlpzx 
1. i moved permissions.py from constants to services and renamed to core_permissions
2. decorators are now the part of services

File: backend/dataall/core/permissions/__init__.py
Comment (dlpzx) [inline backend/dataall/core/permissions/__init__.py:1]: Why are we removing this import? I have not checked if it is explicitly necessary but i would assume it is

File: backend/dataall/core/permissions/db/__init__.py
Comment (dlpzx) [inline backend/dataall/core/permissions/db/__init__.py:1]: How is this operation handled now? Is there any impact for newly deployed data.all infra?

**Comment by dlpzx** [comment]
Hi @SofiaSazonova thanks for updating the PR. How have you tested that everything works as expected?

This PR is pretty packed, but I think there are 2 improvements that we should implement once it is merged:
1. separate `core_permissions` into `environment_permissions` , `organization_permissions`... and include them as part of their core module
2. improve the validator of the resolvers. Right now there are a bunch of functions that I think we could condense in less code. This affects not only these modules in core but also other modules. I was thinking of some utility in `utils` that validates input parameters in a more abstract way. What are your thoughts?

File: backend/dataall/core/permissions/db/__init__.py
Comment (SofiaSazonova) [reply backend/dataall/core/permissions/db/__init__.py:1]: I moved the definition to the new class `TenantPolicyService`, as it seems to belong there.

This function is called in `backend/api_handler.py` and `backend/local_graphql_server.py` as it was before.

File: backend/dataall/core/permissions/__init__.py
Comment (SofiaSazonova) [reply backend/dataall/core/permissions/__init__.py:1]: Hm. It's in fact necessary for the other imports. So I better bring it back.

**Comment by dlpzx** [comment]
It looks good, the only thing is the part of testing. Could you include how have you tested this PR?

**Comment by SofiaSazonova** [comment]
@dlpzx please have a look. The last commit is better to be reviewed separately as there are a lot of moving of files there

File: backend/dataall/core/environment/api/resolvers.py
Comment (dlpzx) [inline backend/dataall/core/environment/api/resolvers.py:31]: We should place the permissions of a module inside the module itself as we do in any other module. For example, take notebooks, we define `notebook_permissions` in `modules/notebooks/service/notebooks_permissions.py`. In the same way environment permissions should be defined in `core/environments/service/environment_permissions.py` And the same applies to other core modules in this last commit

**Review by dlpzx — CHANGES_REQUESTED** [review]
Check comment

**Comment by dlpzx** [comment]
Morning @SofiaSazonova! Thanks for adding the tests :) :) 
One last comment, I suggested moving the permissions to their own core modules, but the permissions seem to be all in the `permissions` module:
image

Instead of `core/permissions/service/environment_permissions.py` ----> `core/environments/service/environment_permissions.py`, wdyt?

**Comment by dlpzx** [comment]
> @dlpzx Organizations, groups and environments permissions can be moved to its own module, but tenants, resources and networks not, so IMHO it will be confusing: half of permissions here, half there.

So the thing is that if I am developing something related to `environments` I would want to modify code in the `environments` module only. The only 2 permissions that are not part of a core/module are `resources_permissions` and `tenant_permissions` because `network_permissions` are used in the `core/vpc` module.

`resources_permissions` does not define any new permission, it defines the set of permissions `RESOURCES_ALL`. In the same way we should move some of the tenant permissions to their respective modules (e.g. `MANAGE_ORGANIZATIONS` to organizations) and `tenant_permissions` would define the `TENANT_ALL` only.

**Review by dlpzx — APPROVED** [review]
Looks good. Approving, @SofiaSazonova clarified offline that moving permissions around was causing circular dependencies. If we want to "modularize" core modules further we can do it in a separate PR

**Pull request merged**
